### PR TITLE
Upgrade `k8s.io/*` to `v0.29.2`, `sigs.k8s.io/controller-runtime` to `v0.17.2`

### PR DESCRIPTION
**Pull request by shafeeqes**
**How to categorize this PR?**

/area open-source dev-productivity
/kind enhancement

**What this PR does / why we need it**:
Upgrade `k8s.io/*` to `v0.29.2`, `sigs.k8s.io/controller-runtime` to `v0.17.2`

Notable changes:
- Adapt the new [restmapper](https://github.com/kubernetes-sigs/controller-runtime/blob/main/pkg/client/apiutil/restmapper.go) implementation.
https://github.com/kubernetes-sigs/controller-runtime/pull/2571. Follow up of https://github.com/gardener/gardener/issues/8382.
- Replace `ResourceRequirements` struct with `VolumeResourceRequirements` for use with volumes: https://github.com/kubernetes/kubernetes/pull/118653
- Fake client now sets TypeMeta only for unstructured objects: https://github.com/kubernetes-sigs/controller-runtime/pull/2633

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener/issues/9047

**Special notes for your reviewer**:
/hold
In draft until https://github.com/gardener/gardener/pull/8989 is merged.

**Release note**:

```other dependency
The following dependencies are updated:
- `k8s.io/*` : `v0.28.4` -> `v0.29.2`
- `sigs.k8s.io/controller-runtime`: `v0.16.3` -> `v0.17.2`
```

### CODE REVIEW

**Comment by shafeeqes** [comment]
Updating `kube-openapi` to `582cce7` since the latest HEAD contains lot of breaking changes, will update it separately

File: pkg/client/kubernetes/cache/single_object.go [hunk 1]
```diff
@@ -148,6 +148,10 @@ func (s *singleObject) GetInformer(ctx context.Context, obj client.Object, opts
 	return cache.GetInformer(ctx, obj, opts...)
 }
 
+func (s *singleObject) RemoveInformer(_ context.Context, _ client.Object) error {
+	return nil
```
Comment (rfranzke) [inline pkg/client/kubernetes/cache/single_object.go]: Why is this not implemented? Can't we implement it in the same way like `GetInformer`?

File: cmd/gardener-resource-manager/app/app.go [hunk 2]
```diff
@@ -189,13 +188,9 @@ func run(ctx context.Context, log logr.Logger, cfg *config.ResourceManagerConfig
...
-				// TODO(ary1992): The new rest mapper implementation doesn't return a NoKindMatchError but a ErrGroupDiscoveryFailed
-				// when an API GroupVersion is not present in the cluster. Remove the old restmapper usage once the upstream issue
-				// (https://github.com/kubernetes-sigs/controller-runtime/pull/2425) is fixed.
-				return thirdpartyapiutil.NewDynamicRESTMapper(
+				return apiutil.NewDynamicRESTMapper(
 					config,
-					thirdpartyapiutil.WithLazyDiscovery,
-					thirdpartyapiutil.WithLimiter(rate.NewLimiter(rate.Every(1*time.Minute), 1)), // rediscover at maximum every minute
+					httpClient,
 				)
```
Comment (rfranzke) [inline cmd/gardener-resource-manager/app/app.go]: Wouldn't this work as well?
```suggestion
			opts.MapperProvider = apiutil.NewDynamicRESTMapper
```

File: cmd/gardenlet/app/app.go [hunk 3]
```diff
@@ -170,12 +170,9 @@ func run(ctx context.Context, cancel context.CancelFunc, log logr.Logger, cfg *c
...
 		MapperProvider: func(config *rest.Config, httpClient *http.Client) (meta.RESTMapper, error) {
-			// TODO(ary1992): The new rest mapper implementation doesn't return a NoKindMatchError but a ErrGroupDiscoveryFailed
-			// when an API GroupVersion is not present in the cluster. Remove the old restmapper usage once the upstream issue
-			// (https://github.com/kubernetes-sigs/controller-runtime/pull/2425) is fixed.
-			return thirdpartyapiutil.NewDynamicRESTMapper(
+			return apiutil.NewDynamicRESTMapper(
 				config,
-				thirdpartyapiutil.WithLazyDiscovery,
+				httpClient,
 			)
```
Comment (rfranzke) [inline cmd/gardenlet/app/app.go]: same here?

File: cmd/gardenlet/app/app.go [hunk 4]
```diff
@@ -344,12 +341,9 @@ func (g *garden) Start(ctx context.Context) error {
...
 		opts.MapperProvider = func(config *rest.Config, httpClient *http.Client) (meta.RESTMapper, error) {
-			// TODO(ary1992): The new rest mapper implementation doesn't return a NoKindMatchError but a ErrGroupDiscoveryFailed
-			// when an API GroupVersion is not present in the cluster. Remove the old restmapper usage once the upstream issue
-			// (https://github.com/kubernetes-sigs/controller-runtime/pull/2425) is fixed.
-			return thirdpartyapiutil.NewDynamicRESTMapper(
+			return apiutil.NewDynamicRESTMapper(
 				config,
-				thirdpartyapiutil.WithLazyDiscovery,
+				httpClient,
 			)
```
Comment (rfranzke) [inline cmd/gardenlet/app/app.go]: same here?

File: test/integration/gardenlet/seed/care/care_test.go [hunk 5]
```diff
@@ -84,9 +80,6 @@ var _ = Describe("Seed Care controller tests", func() {
 					},
 				},
 			},
-			MapperProvider: func(config *rest.Config, httpClient *http.Client) (meta.RESTMapper, error) {
```
Comment (rfranzke) [inline test/integration/gardenlet/seed/care/care_test.go]: Why do we no longer need to set it here?

**Review by rfranzke — COMMENTED** [review]
Yet another great, well-structured PR. I really enjoy reviewing them. Thank you!

**Comment by shafeeqes** [comment]
/hold 
Due to an internal discussion with @ary1992

**Pull request merged**